### PR TITLE
Add Go example for TinyTodo based on cedar-go

### PR DESCRIPTION
**Pull request by mqf20**
Also see https://github.com/cedar-policy/cedar-go/issues/24.

### CODE REVIEW

**Comment by mqf20** [comment]
Hey @andrewmwells-amazon, I've prepared the PR as suggested.

There seems to be a problem with DCO sign off, do you need me to sign or does it only concern AWS staff?

**Comment by khieta** [comment]
Hi @mqf20! To pass the DCO check you should include `-s` in your git commit commands. I believe if you click the "Details" link next to the CI failure it'll take you to a page that explains more.

**Comment by khieta** [comment]
(Also: thanks for the contribution! I'll tag some reviewers who have familiarity with Go & the TinyTodo app)

**Comment by mqf20** [comment]
Got it, I just wasn't sure if the signing policy applies only to AWS staff.

Should I withdraw this PR and resubmit a new one with a signed commit?

File: tinytodo-go/entities.json [hunk 1]
```diff
@@ -0,0 +1,66 @@
+{
```
Comment (shaobo-he-aws) [inline tinytodo-go/entities.json:1]: If this file is exactly the same as the Rust version, can we use one copy instead?

File: tinytodo-go/test_tinytodo.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
+from tinytodo import *
```
Comment (shaobo-he-aws) [inline tinytodo-go/test_tinytodo.py:1]: I assume that the go version uses the same "frontend" Python file as the Rust version.. If so, we should also use one copy only.

**Comment by shaobo-he-aws** [comment]
A high-level feedback from me is to reuse the files in the existing TinyTodo example (e.g., policies, entities, and Python files) as much as possible, which makes me wonder if we should restructure the TinyTodo folder into something like,


tinytodo
  - policy.cedar
  - tinytodo.cedarschema
  - tinytodo.py
  - ...
  - /rust
  - /go

File: tinytodo-go/entities.json (hunk 1 above)
Comment (mqf20) [reply tinytodo-go/entities.json:1]: Agreed, this is what I was going for

**Comment by mqf20** [comment]
> A high-level feedback from me is to reuse the files in the existing TinyTodo example (e.g., policies, entities, and Python files) as much as possible, which makes me wonder if we should restructure the TinyTodo folder into something like,
> 
> ```
> tinytodo
>   - policy.cedar
>   - tinytodo.cedarschema
>   - tinytodo.py
>   - ...
>   - rust
>     - Cargo.toml
>     - src
>   - go
>     - ...
> ```

Agreed, this sounds neater.

However, the Go implementation does not support some features (templates, schemas), so the Rust and Go implementations are not equivalent.

File: tinytodo-go/internal/app/server/context_test.go [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+	"encoding/json"
+	"github.com/cedar-policy/cedar-go"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"os"
+	"path"
+	"testing"
+)
+
+func readFile(t *testing.T, prefix, filename string) []byte {
```
Comment (patjakdev) [inline tinytodo-go/internal/app/server/context_test.go:16]: Hmmm, I'm not really understanding the need for prefix and filename in this helper. It seems like you could just pass a single argument.

File: tinytodo-go/internal/app/server/entitystore/list.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import (
+	"code.byted.org/binaryauthorization/tinytodo-go/internal/app/server/entitystore/entitytype"
+	"code.byted.org/binaryauthorization/tinytodo-go/internal/app/server/entitystore/taskstate"
+	"github.com/cedar-policy/cedar-go"
+	"strconv"
+)
+
+// ListUID is a transparent wrapper around EntityUID, to make it clear that we want a List's EntityUID.
+//
+// Note that we use inheritance instead of alias, because we want to inherit methods. See [blog post].
```
Comment (patjakdev) [inline tinytodo-go/internal/app/server/entitystore/list.go:12]: Tiny nit: this is "composition" or "embedding" rather than "inheritance".

File: tinytodo-go/tinytodo.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+        else:
+            raise Exception('Expected type: %s, got %s' % (expected_type, typ))
+    else:
+        raise Exception('Failed to parse euid: {}' % euid)
+
+
+def attempt_build():
+    print('Attempting to build the tiny-todo-server via Go')
+    proc = subprocess.run(['go', 'build', '-o', server_binary_path, './cmd/server'], check = True)
+    print('Build successfull')
```
Comment (patjakdev) [inline tinytodo-go/tinytodo.py:74]: tiny nit: `successful`

File: tinytodo-go/test_tinytodo.py (hunk 2 above)
Comment (patjakdev) [reply tinytodo-go/test_tinytodo.py:1]: Agreed, although it's not totally clear to me what value the Python frontend adds over just running `go build` and `cmd/server` directly. Maybe it lets you interact with the server from a REPL somewhat easily? Dunno.

**Review by patjakdev — CHANGES_REQUESTED** [review]
Thank you so much for writing all of this up and adeptly skirting some of the limitations of the current `cedar-go` implementation!

I mostly skimmed over the actual application logic, but I did look more closely at the interface with `cedar-go` and had a few questions. I decided to request changes because I think it's pretty important that `isAuthorized()` adhere to the Cedar ethos of respecting the policy decision, even in the face of errors.

File: tinytodo-go/entities.json (hunk 1 above)
Comment (mqf20) [reply tinytodo-go/entities.json:1]: Before proceeding on this, perhaps the maintainers of this repository should decide whether we wish to go with @shaobo-he-aws's [suggestion](https://github.com/cedar-policy/cedar-examples/pull/182#issuecomment-2269883747) to restructure the entire TinyTodo example?

File: tinytodo-go/internal/app/server/context_test.go (hunk 3 above)
Comment (mqf20) [reply tinytodo-go/internal/app/server/context_test.go:16]: Will change this

File: tinytodo-go/internal/app/server/entitystore/entityuid.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+import (
+	"code.byted.org/binaryauthorization/tinytodo-go/internal/app/server/entitystore/entitytype"
+	"encoding/json"
+	"fmt"
+	"github.com/cedar-policy/cedar-go"
+	"strings"
+)
+
+// EntityUID is a transparent wrapper around cedar.EntityUID used to represent entities in EntityStore, so that we
+// can define our own UnmarshalJSON and MarshalJSON functions.
```
[the remark this answers is not in the document]
Comment (mqf20) [reply tinytodo-go/internal/app/server/entitystore/entityuid.go:12]: I had to write my own `UnmarshalJSON` method for `EntityUID` (as well as `UserUID`, `ListUID`, `TeamUID` by extension) because I couldn't unmarshal `entities.json`.

Take the user `kesha` as an example:

```json
{
  "users": {
    "User::\"kesha\"": {
      "euid": "User::\"kesha\"",

      // omitted
```

`kesha`'s UID would be `User::\"kesha\"`, and I wanted to unmarshal all users from `entities.json` into a `map[UserUID]*User`. If I did not write my own `UnmarshalJSON` method, I would rely on`cedar.EntityUID`'s `UnmarshalJSON` method which could only unmarshal `{"type":"User","id":"kesha"}`.

In short, `entities.json` does not adhere to [Cedar's Entities and Context syntax](https://docs.cedarpolicy.com/auth/entities-syntax.html), which is why I had to write my own `UnmarshalJSON` method.

I had to write my own `MarshalJSON` method because the Python client (`tinytodo.py`) expects the returned values to be in the format of `User::\"kesha\"` rather than `{"type":"User","id":"kesha"}`.

(I have a feeling that I may have missed an obvious solution or done something that's not idiomatic in Go.)

File: tinytodo-go/internal/app/server/entitystore/list.go (hunk 4 above)
Comment (mqf20) [reply tinytodo-go/internal/app/server/entitystore/list.go:12]: Thanks for spotting the mistake, will change this

File: tinytodo-go/internal/app/server/context.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+			fmt.Sprintf("Policy: %d", r.Policy),
+		)
+	}
+
+	if len(diag.Errors) > 0 {
+		var errs []error
+		for _, e := range diag.Errors {
+			errs = append(errs, errors.New(e.String()))
+		}
+		return false, nil, errors.Join(errs...)
```
[the remark this answers is not in the document]
Comment (mqf20) [reply tinytodo-go/internal/app/server/context.go:52]: You're right, the Go implementation does not match the [Rust implementation](https://github.com/cedar-policy/cedar-examples/blob/d70996e4ebc78fa41f4872e9bfb69b457aa59021/tinytodo/src/context.rs#L565.

Before I fix the Go implementation, please correct me if I have misinterpreted the [Rust implementation](https://github.com/cedar-policy/cedar-examples/blob/d70996e4ebc78fa41f4872e9bfb69b457aa59021/tinytodo/src/context.rs#L565) (I'm new to Rust):

```rust
        let response = self.authorizer.is_authorized(&q, &self.policies, &es);
        info!("Auth response: {:?}", response);
        match response.decision() {
            Decision::Allow => Ok(()),
            Decision::Deny => Err(Error::AuthDenied(response.diagnostics().clone())),
        }
```

- Returns an empty successful value if the decision is *allow*
- Returns an error with the diagnostics (errors and reasons) if the decision is *deny*

Based on Cedar's skip-on-error behavior that you quoted, it seems like the Rust implementation has overlooked a scenario: the decision is *allow* but errors are present.

File: tinytodo-go/entities.json (hunk 1 above)
Comment (shaobo-he-aws) [reply tinytodo-go/entities.json:1]: It seems that no maintainers have rejected my proposal. I take it as a green light.

File: tinytodo-go/test_tinytodo.py (hunk 2 above)
Comment (shaobo-he-aws) [reply tinytodo-go/test_tinytodo.py:1]: Yes. We use the Python REPL to interact with the backend server.

File: tinytodo-go/tinytodo.py (hunk 5 above)
Comment (shaobo-he-aws) [reply tinytodo-go/tinytodo.py:74]: Please go ahead.

File: tinytodo-go/entities.json (hunk 1 above)
Comment (mqf20) [reply tinytodo-go/entities.json:1]: Perhaps we should implement @shaobo-he-aws's suggestion to restructure the entire TinyTodo example in a follow-up PR?

Comment (shaobo-he-aws) [reply tinytodo-go/entities.json:1]: Either way works for me.

Comment (patjakdev) [reply tinytodo-go/entities.json:1]: Restructuring in a follow-up sounds good to me.

File: tinytodo-go/internal/app/server/context.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (patjakdev) [reply tinytodo-go/internal/app/server/context.go:52]: > Based on Cedar's skip-on-error behavior that you quoted, it seems like the Rust implementation has overlooked a scenario: the decision is allow but errors are present.

I think the Rust implementation is following the spirit of the skip-on-error behavior. If the decision is allow, the request should be authorized, even if there are errors. Those errors are logged via the `info!()` macro above the `match` statement.

File: tinytodo-go/internal/app/server/entitystore/entityuid.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (patjakdev) [reply tinytodo-go/internal/app/server/entitystore/entityuid.go:12]: I see. Thank you for clarifying. It does seem like `entities.json` should strive to conform to the published syntax. Maybe that can be addressed in a later PR that merges the two TinyTodo examples.

File: tinytodo-go/entities.json (hunk 1 above)
Comment (mqf20) [reply tinytodo-go/entities.json:1]: I've created a new issue to implement this suggestion.

File: tinytodo-go/internal/app/server/context.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mqf20) [reply tinytodo-go/internal/app/server/context.go]: Got it, let me know if [this commit](https://github.com/cedar-policy/cedar-examples/pull/182/commits/5ad4eebd2b717c2e1c0caacdf8b92e7fc438f1e9) is aligned with the Rust implementation

File: tinytodo-go/internal/app/server/context.go [hunk 8]
```diff
@@ -14,18 +15,19 @@ import (
 //
 // Non-existent entities (resources) will result in an error. (TODO: we may not want this behaviour)
 func (s *Server) isAuthorized(
+	ctx context.Context,
```
Comment (patjakdev) [inline tinytodo-go/internal/app/server/context.go:18]: nit: I'd consider putting this argument last to match the normal PARC order.

**Review by patjakdev — APPROVED** [review]
Looks good! Thanks for taking my feedback.

File: tinytodo-go/internal/app/server/context.go (hunk 8 above)
Comment (mqf20) [reply tinytodo-go/internal/app/server/context.go:18]: In this case, the context here is different from the [context in an authorization request (the C in PARC)](https://pkg.go.dev/github.com/cedar-policy/cedar-go#Request). In the Rust implementation, the authorization request is also missing.

Perhaps I should make that clear?

**Comment by mqf20** [comment]
@patjakdev just pushed one more commit 9839aea601b0de587b0d561c8608a3646fce8568 to clean up the Go module paths

**Comment by mqf20** [comment]
@aaronjeline @shaobo-he-aws do you have any comments on this PR?

**Review by shaobo-he-aws — APPROVED** [review]
LGTM, as a person who rarely knows Go.

**Review by patjakdev — APPROVED** [review]

File: tinytodo-go/internal/app/server/httphandlersapi.go [hunk 9]
```diff
@@ -0,0 +1,1493 @@
...
+	UID  string `json:"uid"`
+	Name string `json:"name"`
+}
+
+// handlePostAPIListCreate handles POST requests to "/api/list/create".
+//
+// Example response:
+//
+//	List::"0"
+func (s *Server) handlePostAPIListCreate(w http.ResponseWriter, r *http.Request) {
```
Comment (aaronjeline) [inline tinytodo-go/internal/app/server/httphandlersapi.go:137]: Looks like there's ~50 lines of calling Cedar at each auth endpoint. It might be a better demo if we can abstract that somehow? Also helps with maintenance and ensuring you don't forget to call Cedar

**Review by aaronjeline — APPROVED** [review]
Looks good overall. Two minor requests:
1) The above comment, if there isn't a clear solution it's fine as is.
2) Can you adjust the CI to auto-build this?

**Review by aaronjeline — APPROVED** [review]
Looks good, once CI passes :)
